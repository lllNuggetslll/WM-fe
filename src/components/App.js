import React, { Component } from "react";
import { object, bool, func } from "prop-types";
import { connect } from "react-redux";
import { locate, fetchListing } from "../actions/actions";
import logo from "../assets/logo.png";
import { Modal } from "./modal";
import Loader from "./loader";
import { setParams, getParams } from "./../utils/urlUtils";
import HeroSection from "./hero_section";
import AppContent from "./app_content";
import styled from "styled-components";

export const AppHeader = styled.div`
  min-height: 70px;
  display: flex;
  padding: 0 20px;
  justify-content: space-between;
  align-items: center;
  background-color: #222;
  color: #fff;

  img {
    width: 110px;
    height: 25px;
  }
`;

export const AppWrapper = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  min-width: 520px;
  margin-bottom: 20px;
  text-align: center;
`;

export class App extends Component {
  componentDidMount() {
    const {
      locate,
      fetchListing,
      history: {
        location: { search }
      },
      getParams
    } = this.props;
    const loc = getParams("location", search);
    const listing = getParams("listing", search);

    if (loc) {
      const [latitude, longitude] = loc.split(",");

      locate({ latitude, longitude });
    }

    if (listing) {
      fetchListing(listing);
    }
  }

  locateMe = () => {
    const { locate, setParams } = this.props;

    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(position => {
        const { latitude, longitude } = position.coords;
        const url = setParams({
          param: "location",
          query: `${latitude},${longitude}`
        });
        locate(position.coords);

        this.props.history.push(`?${url}`);
      });
    }
  };

  render() {
    const { isLocating, location, regions, error, fetchListing } = this.props;

    return (
      <AppWrapper>
        <Modal />
        <Loader isLoaded={!isLocating} />
        <AppHeader>
          <img src={logo} alt="weedmaps logo" />
        </AppHeader>
        <HeroSection {...{ location, isLocating }} locateMe={this.locateMe} />
        <AppContent {...{ regions, error, fetchListing }} />
      </AppWrapper>
    );
  }
}

App.defaultProps = {
  // for testing, so i can pass in mock funcs
  getParams,
  setParams,
  isLocating: false,
  location: {},
  regions: {},
  error: {}
};

App.propTypes = {
  isLocating: bool.isRequired,
  location: object,
  regions: object,
  error: object,
  locate: func.isRequired,
  fetchListing: func.isRequired,
  history: object.isRequired,
  getParams: func.isRequired,
  setParams: func.isRequired
};

const mapStateToProps = state => state.location;
// For testing without redux
export { App as TestApp };

export default connect(mapStateToProps, { locate, fetchListing })(App);
