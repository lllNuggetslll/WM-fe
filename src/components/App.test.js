import React from "react";
import { shallow } from "enzyme";
import { TestApp as App } from "./App";

const defaultProps = {
  isLocating: false,
  location: {},
  regions: {},
  error: {},
  locate: () => {},
  fetchListing: () => {},
  history: {
    push: () => {},
    location: {
      search:
        "?location=34.035243900000005%2C-118.25220709999999&listing=335715637"
    }
  },
  getParams: () => {}
};

describe("<App/>", () => {
  it("should render App", () => {
    const wrapper = shallow(<App {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe("componentDidMount", () => {
    it("should fire getParams twice", () => {
      const getParams = jest.fn();
      shallow(<App {...defaultProps} {...{ getParams }} />);

      expect(getParams.mock.calls.length).toEqual(2);
    });

    it("should fire locate if there is a location query in the url", () => {
      const locate = jest.fn();
      const getParams = () => "lat,lng";
      shallow(<App {...defaultProps} {...{ locate, getParams }} />);

      expect(locate).toHaveBeenCalled();
      expect(locate).toHaveBeenCalledWith({
        latitude: "lat",
        longitude: "lng"
      });
    });

    it("should fire fetchListing if there is a listing query in the url", () => {
      const fetchListing = jest.fn();
      const getParams = () => "listingQuery";
      shallow(<App {...defaultProps} {...{ fetchListing, getParams }} />);

      expect(fetchListing).toHaveBeenCalled();
      expect(fetchListing).toHaveBeenCalledWith(getParams());
    });
  });

  describe("locateMe", () => {
    it("should fire navigator, setParams, locate, and history.push ", () => {
      const setParams = jest.fn();
      const locate = jest.fn();
      const history = {
        push: jest.fn(),
        location: {
          search: ""
        }
      };
      const wrapper = shallow(
        <App {...defaultProps} {...{ setParams, locate, history }} />
      );

      wrapper.instance().locateMe();

      expect(
        global.navigator.geolocation.getCurrentPosition
      ).toHaveBeenCalled();
      expect(setParams).toHaveBeenCalled();
      expect(locate).toHaveBeenCalled();
      expect(history.push).toHaveBeenCalled();
    });
  });
});
