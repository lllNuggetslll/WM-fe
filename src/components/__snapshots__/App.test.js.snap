// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> should render App 1`] = `
<styled.div>
  <Connect(withRouter(DisplayModal)) />
  <Loader
    isLoaded={true}
  />
  <styled.div>
    <img
      alt="weedmaps logo"
      src="logo.png"
    />
  </styled.div>
  <Hero
    isLocating={false}
    locateMe={[Function]}
    location={Object {}}
  />
  <Content
    error={Object {}}
    fetchListing={[Function]}
    regions={Object {}}
  />
</styled.div>
`;
